### PR TITLE
Improve js_of_ocaml support

### LINKED ISSUE #970: Make jsoo separate compilation independent of dev
Discussing this with @jordwalke, we've reached the conclusion that separate compilation should really be its own independent option that you can toggle for any profile.

Actually, I'm thinking if it's best to have this option env scoped as well. It seems like we should be able to mix and match this stuff within a single project.

I still think that this option should be the default exactly as it stands today (as `--dev` is default) but perhaps Jordan some arguments against this.

### LINKED ISSUE #1613: Per environment flags with js_of_ocaml
I want to use different js_of_ocaml flags in my project, depending on whether we are in dev or release profile.

Dune [already adds](https://github.com/ocaml/dune/blob/55531596991f0f2d3b60f24183ea0eb3f80c2fd6/src/js_of_ocaml_rules.ml#L12) `--pretty` and `--source-maps-inline` in dev, however, there are more jsoo options to play with, and I get dramatically better debugging with the [recommended settings](https://ocsigen.org/js_of_ocaml/dev/manual/options):

```
For Debugging: "–pretty –no-inline –debug-info" + eventually "–disable staticeval –disable share"
For Production: "–opt 3". It minimize the generated javascript by applying various optimizations until a fix-point is reached
```

I tried using an `env` block, but couldn't make it work any way I tried it. Is there an existing way to do this, and if not:
- can a way to do this be added?
- can we pick better defaults for the dev and production env?

### LINKED ISSUE #4027: when building js_of_ocaml program: "while linking cmo, the external function [...] is not available"
## Expected Behavior

```lisp
(executable
 (name foo)
 (modes js)
 (libraries bonsai.web core_kernel)
 (js_of_ocaml (flags --no-sourcemap) (javascript_files example.js))
 (preprocess (pps ppx_jane js_of_ocaml-ppx)))
```

In this executable, which is for , I link "example.js" which defines some external functions.  In the ocaml code, I use 

```ocaml
external my_extern_fun: (float Js.js_array Js.t -> float) -> unit = "my_extern_fun"
```

I expect this program to build.

## Actual Behavior

The build fails with the following 
```
$ dune build @default
File "_none_", line 1: 
Error: Error while linking ui/.pixelmapper.eobjs/byte/dune__exe__Pixelmapper.cmo:
       The external function `my_extern_fun' is not available
```

If I ask for just the .bc.js file, there is no error

```
$ dune build ui/pixel_mapper.bc.js
Done: 0/0 (jobs: 0)
```
and the file is built correctly.

The fact that @default is doing something _extra_ that fails is quite confusing.

## Specifications

- Version of `dune` (output of `dune --version`): 2.7.1
- Version of `ocaml` (output of `ocamlc --version`): 4.10.1

### LINKED ISSUE #4444: dune-build-info only works when public_name is specified, even if only mode is JS
## Expected Behavior
I was expecting to be able to include `dune-build-info` (specifically the project version) in an executable target that only supports the `js` mode.

For example:
```
(executable
 (name Example)
 (modes js)
 (modules Example)
 (libraries dune-build-info))
```

## Actual Behavior
`Build_info.V1.version ()` is always `None` because the executable doesn't specify a `public_name`, but adding a `public_name` if I'm only specifying `(modes js)` results in this error:

```
Error: No installable mode found for this executable.
When public_name is set, one of the following modes is required:
- exe
- native
- byte
```

## Reproduction

1. Clone https://github.com/phated/dune-build-info-jsoo
1. Run `esy` inside the repo
1. Run `esy test`

Adding an `exe` mode and `public_name` to that executable allows the JS to include the project version.

## Specifications

- Version of `dune` (output of `dune --version`): 2.8.5
- Version of `ocaml` (output of `ocamlc --version`): 4.11.0
- Operating system (distribution and version): MacOS 10.15 (Catalina)

### LINKED ISSUE #4951: Make the tests stanza setup runtest aliases for js
## Desired Behavior
```
(tests
  (names a b)
  (modes native js))
```
And `dune runtest` should/could run js tests (similar to inline_test with js mode)

### LINKED ISSUE #4999: Allow to customize runtest aliases for inline_tests and tests stanza
## Desired Behavior

I'd like to be able to customize the runtest alias when using `(inline_tests)` in a library stanza or `(tests ..)`.
I'd like to customize this alias per modes.
The goal is to run js tests only when node is installed / available.

## Example

```
(library
  (name lib)
  (inline_tests (modes (js (alias runtest-js)) (native (alias runtest-native)))))
```

```
(tests
   (names a b c)
   (modes (js (alias runtest-js)) (native (alias runtest-native))))
```
related to #4951

```
dune build @runtest-js
dune build @runtest-native
```

### PR DESCRIPTION
**Pull request by hhugo**
Please review commit by commit.

- [x] the tests stanza should run tests for all the modes defined. Allowing running tests in javascript automatically (fix #4951)
- [x] configure the default alias for runtest with js_of_ocaml. The main motivation is that we want people to be able to run tests even without node installed (fix #4999)
- [x] configure compilation mode in env (fix #970)
- [x] configure jsoo flags in env (fix #1613)
- [x] make js build work with disable_dynamically_linked_foreign_archives=true
- [x] pass "-no-check-prims" when building bytecode for jsoo (fix #4027)
- [x] add/fix support for build-info+jsoo (fix #4444)
- [x] add tests
- [x] udpate documentation

Extends https://github.com/ocaml/dune/pull/1044

### CODE REVIEW

File: src/dune_rules/test_rules.ml [hunk 1]
```diff
@@ -17,58 +17,93 @@ let rules (t : Dune_file.Tests.t) ~sctx ~dir ~scope ~expander ~dir_contents =
       `Regular
   in
   let open Memo.Build.O in
+  let modes =
+    if Dune_project.dune_version (Scope.project scope) < (3, 0) then
+      [ (Dune_file.Executables.Link_mode.exe, Loc.none) ]
```
[the remark this answers is not in the document]
Comment (hhugo) [reply src/dune_rules/test_rules.ml]: I refactored the code a bit, not needed anymore

**Comment by hhugo** [comment]
@nojb, I'm not planning to fix #4213 in the PR. sorry.

**Comment by hhugo** [comment]
@rgrinberg, I think this is ready for another round of review !

**Comment by hhugo** [comment]
@nojb, I changed by mind, the PR also fixes #4213

File: test/blackbox-tests/test-cases/jsoo/tests.t/run.t [hunk 2]
```diff
@@ -1,8 +1,6 @@
 tests stanza with jsoo
 
-  $ dune runtest
-  a: ok
-  b: ok
-  File "run.t", line 1, characters 0-0:
-  Error: Files _build/default/run.t and _build/default/run.t.corrected differ.
+  $ dune build @runtest-js
+  Error: Alias "runtest-js" specified on the command line is empty.
```
Comment (hhugo) [inline test/blackbox-tests/test-cases/jsoo/tests.t/run.t:4]: @rgrinberg, @jeremiedimino how can I convince dune that this alias exists ?

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/jsoo/tests.t/run.t:4]: You can make it a "standard" alias. See function `make_standard` in `src/dune_engine/alias.ml`.

Comment (hhugo) [reply test/blackbox-tests/test-cases/jsoo/tests.t/run.t:4]: I want to have the alias configurable in env by users,  does it make sense to call make_standard ?

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/jsoo/tests.t/run.t:4]: Standard only means that if the alias is empty then Dune doesn't complain about it.

**Comment by hhugo** [comment]
I don't understand some diffs when running tests. Any idea ?
```diff
$ make test
./dune.exe runtest
File "test/blackbox-tests/test-cases/exec-missing.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/exec-missing.t/run.t
++++++ test/blackbox-tests/test-cases/exec-missing.t/run.t.corrected
File "test/blackbox-tests/test-cases/exec-missing.t/run.t", line 8, characters 0-1:
 |When using dune exec, the external-lib-deps command refers to the executable:
 |
 |  $ dune exec ./x.exe
 |  File "dune", line 3, characters 12-26:
 |  3 |  (libraries does-not-exist))
 |                  ^^^^^^^^^^^^^^
 |  Error: Library "does-not-exist" not found.
-|  -> required by _build/default/x.exe
 |  [1]
File "test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t
++++++ test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t.corrected
File "test/blackbox-tests/test-cases/enabled_if/eif-simple.t/run.t", line 5, characters 0-1:
 |Test that `enabled_if` fields work as expected.
 |
 |This alias is disabled, building it should do nothing:
 |  $ dune build @x
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
+|  [1]
 |
 |This one is enabled:
 |  $ dune build @y
-|  Building alias y
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
+|  [1]
 |
 |This rule is disabled, trying to build a should fail:
 |  $ dune build a
-|  Error: Don't know how to build a
-|  Hint: did you mean b?
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
 |  [1]
 |
 |This one is enabled:
 |  $ dune build b
-|  Building file b
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
+|  [1]
 |
 |Test the enabled_if field for libraries:
 |
 |  $ dune build foo
-|  Error: Don't know how to build foo
+|  File "dune", line 35, characters 12-15:
+|  35 |  (libraries foo))
+|                   ^^^
+|  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
+|  -> required by library "bar" in _build/default
+|  -> required by executable main in dune:44
 |  [1]
 |
 |  $ dune build main.exe
 |  File "dune", line 35, characters 12-15:
 |  35 |  (libraries foo))
 |                   ^^^
 |  Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
 |  -> required by library "bar" in _build/default
 |  -> required by executable main in dune:44
-|  -> required by _build/default/main.exe
 |  [1]
File "test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t
++++++ test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t.corrected
File "test/blackbox-tests/test-cases/virtual-libraries/double-implementation.t/run.t", line 10, characters 0-1:
 |Executable that tries to use two implementations for the same virtual lib
 |  $ dune build
 |  Error: Conflicting implementations for virtual library "vlib" in
 |  _build/default/vlib:
 |  - "impl1" in _build/default/impl1
 |    -> required by library "bar" in _build/default
 |  - "impl2" in _build/default/impl2
 |  This cannot work.
 |  -> required by executable foo in dune:2
-|  -> required by _build/default/foo.exe
-|  -> required by alias default in dune:11
 |  [1]
File "test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t", line 1, characters 0-0:
------ test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t
++++++ test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t.corrected
File "test/blackbox-tests/test-cases/virtual-libraries/missing-implementation.t/run.t", line 6, characters 0-1:
 |Executable that tries to build against a virtual library without an implementation
 |  $ dune build
 |  Error: No implementation found for virtual library "vlib" in
 |  _build/default/vlib.
 |  -> required by executable foo in dune:2
-|  -> required by _build/default/foo.exe
-|  -> required by alias default in dune:5
 |  [1]
make: *** [Makefile:75: test] Error 1
```

File: test/blackbox-tests/test-cases/jsoo/tests.t/run.t (hunk 2 above)
Comment (hhugo) [reply test/blackbox-tests/test-cases/jsoo/tests.t/run.t]: Thanks

**Comment by hhugo** [comment]
> I don't understand some diffs when running tests. Any idea ?

Reply to myself, fixed by https://github.com/ocaml/dune/pull/5049/commits/64517a008c00aea15e529e19f2334145722696a3

File: src/dune_rules/exe.ml [hunk 3]
```diff
@@ -21,7 +21,11 @@ module Linkage = struct
 
   let byte = { mode = Byte; ext = ".bc"; flags = [] }
 
-  let byte_for_jsoo = { mode = Byte_for_jsoo; ext = ".bc-for-jsoo"; flags = [] }
+  let byte_for_jsoo =
+    { mode = Byte_for_jsoo
+    ; ext = ".bc-for-jsoo"
+    ; flags = [ "-no-check-prims" ]
```
Comment (hhugo) [inline src/dune_rules/exe.ml:27]: Should this be guarded by a check on version ?

**Comment by hhugo** [comment]
I've cleaned up the history. It should be fully ready for review.

**Comment by hhugo** [comment]
I'm not sure what to make of the macos failure
```diff
diff --git a/_build/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml b/_build/.sandbox/61b2db98d47d01466544d55fcd57622a/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml.corrected
index c049768..b33149d 100644
--- a/_build/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml
+++ b/_build/.sandbox/61b2db98d47d01466544d55fcd57622a/default/test/expect-tests/dune_file_watcher/dune_file_watcher_tests_macos.ml.corrected
@@ -41,12 +41,13 @@ let%expect_test _ =
   print_events 1;
   [%expect
     {|
-    { path = In_source_tree "."; kind = "Created" }
-    { path = In_source_tree "x"; kind = "Unknown" }
-    Got more events than expected: expected 1, saw 2 |}];
+    { path = In_build_dir "."; kind = "Created" } |}];
   Unix.rename "x" "y";
   print_events 0;
   [%expect {|
+    { path = In_source_tree "x"; kind = "Unknown" }
+    { path = In_source_tree "y"; kind = "Unknown" }
+    Got more events than expected: expected 0, saw 2
 |}];
   let (_ : _) = Fpath.mkdir_p "d/w" in
   Stdio.Out_channel.write_all "d/w/x" ~data:"x";
@@ -56,10 +57,8 @@ let%expect_test _ =
   print_events 1;
   [%expect
     {|
-    { path = In_source_tree "x"; kind = "Unknown" }
-    { path = In_source_tree "y"; kind = "Unknown" }
     { path = In_source_tree "d"; kind = "Created" }
     { path = In_source_tree "d/w"; kind = "Created" }
     { path = In_source_tree "d/w/x"; kind = "Unknown" }
     { path = In_source_tree "d/w/y"; kind = "Unknown" }
-    Got more events than expected: expected 1, saw 6 |}]
+    Got more events than expected: expected 1, saw 4 |}]
make: *** [test] Error 1
'make test' failed with code 2
```

**Comment by rgrinberg** [comment]
You can ignore the mac error for now. I'm fixing it in another PR.

**Comment by ggreif** [comment]
@hhugo there was a scratch file test/blackbox-tests/test-cases/jsoo/no-check-prim.t/lib/dune~ checked in too, was this intentional?

**Comment by hhugo** [comment]
> @hhugo there was a scratch file test/blackbox-tests/test-cases/jsoo/no-check-prim.t/lib/dune~ checked in too, was this intentional?

No, you can get rid of it

**Comment by ggreif** [comment]
> > @hhugo there was a scratch file test/blackbox-tests/test-cases/jsoo/no-check-prim.t/lib/dune~ checked in too, was this intentional?
> 
> No, you can get rid of it

#6923

**Comment by ggreif** [comment]
@hhugo I have trouble getting this working:
```
File "js/dune", line 22, characters 53-63:
22 |  (js_of_ocaml (flags <flags>)(build_runtime <flags>)(link_flags <flags>))
                                                          ^^^^^^^^^^
Error: Unknown field link_flags
```
I am using `dune` v3.2. The line 22 above is straight from the manual. (Forget the `<flags>` placeholders for now, they are irrelevant.)

**Comment by hhugo** [comment]
> @hhugo I have trouble getting this working:
> 
> ```
> File "js/dune", line 22, characters 53-63:
> 22 |  (js_of_ocaml (flags <flags>)(build_runtime <flags>)(link_flags <flags>))
>                                                           ^^^^^^^^^^
> Error: Unknown field link_flags
> ```
> 
> I am using `dune` v3.2. The line 22 above is straight from the manual. (Forget the `<flags>` placeholders for now, they are irrelevant.)

What `dune lang` are you using ?

**Comment by ggreif** [comment]
Yes, `echo '(lang dune 3.0)' > js/dune-project` seems to solve my problem! Thanks!

**Pull request merged**